File: leetcode/easy/27_remove_element.py
# -----------------------------------------
# My Solution
#
# Time  Complexity: O(nk)
# Space Complexity: O(1)
# -----------------------------------------
# k := time complexity of removing element from list
class Solution:
    def removeElement(self, nums: List[int], val: int) -> int:
        pointer = 0
        while pointer < len(nums):
            if nums[pointer] == val:
                nums.pop(pointer)
            else:
                pointer += 1
        return len(nums)

# -----------------------------------------
# Model Solution
#
# Time  Complexity: O(n)
# Space Complexity: O(1)
# -----------------------------------------
# Ref: https://leetcode.com/problems/remove-element/solution/197569
class Solution:
    def removeElement(self, nums: List[int], val: int) -> int:
        count = 0
        for index in range(len(nums)):
            if nums[index] != val :
                nums[count] = nums[index]
                count += 1
        return count

File: leetcode/easy/27_remove_element.cpp
// -----------------------------------------
// My Solution
//
// Time  Complexity: O(n)
// Space Complexity: O(1)
// -----------------------------------------
// n := nums.size()
class Solution {
public:
    int removeElement(vector<int>& nums, int val) {
        int left_cursor_position  = 0;
        int right_cursor_position = nums.size() - 1;
        while (left_cursor_position <= right_cursor_position) {
            while (right_cursor_position >= left_cursor_position && nums[right_cursor_position] == val) {
                right_cursor_position -= 1;
            }
            if (right_cursor_position < left_cursor_position)
                break;

            if (nums[left_cursor_position] == val)
                swap(nums[left_cursor_position], nums[right_cursor_position]);
            left_cursor_position += 1;
        }

        return left_cursor_position;
    }
};
